qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -12,117 +12,479 @@
+        """The spin number.
 
-    # TODO: Make this much more efficient by working with lists and label indices
-    def add(self, other):
-        """Returns a SpinOp representing the sum of the given operators.
+        Returns:
+            Spin number
         """
-        if isinstance(other, SpinOperator):
-            new_operatorlist = copy.deepcopy(self.operator_list)
-            # If the operators have the same label, simply add coefficients:
-            for operator in new_operatorlist:
-                if other.label == operator.label:
-                    sum_coeff = operator.coeff + other.coeff
-                    operator._coeff = sum_coeff  # set the coeff of sum operator to sum_coeff
-                    # if the new coefficient is zero, remove the operator from the list
-                    if sum_coeff == 0:
-                        new_operatorlist.remove(operator)
-                    return SpinOp(new_operatorlist)
-            new_operatorlist.append(other)
-            return SpinOp(new_operatorlist)
-
-        if isinstance(other, SpinOp):
-            new_operatorlist = copy.deepcopy(self.operator_list)
-            for elem in other.operator_list:
-                new_operatorlist.append(elem)
-                # If the operators have the same label, simply add coefficients:
-                for operator in new_operatorlist[:-1]:
-                    if elem.label == operator.label:
-                        new_operatorlist.pop()
-                        sum_coeff = operator.coeff + elem.coeff
-                        operator._coeff = sum_coeff  # set the coeff of sum operator to sum_coeff
-                        # if the new coefficient is zero, remove the operator from the list
-                        if sum_coeff == 0:
-                            new_operatorlist.remove(operator)
-                        break
-            return SpinOp(new_operatorlist)
-
-        raise TypeError("Unsupported operand type(s) for +: 'SpinOp' and "
-                        "'{}'".format(type(other).__name__))
+        return Fraction(self._dim - 1, 2)
 
-    def compose(self, other):
-        raise NotImplementedError()
-
-    def mul(self, other):
-        """Overloads the multiplication operator `*` for self and other, where other is a
-        number-type.
+    @property
+    def x(self) -> np.ndarray:
+        """A np.ndarray storing the power i of (spin) X operators on the spin system.
+        I.e. [0, 4, 2] corresponds to X_2^0 \\otimes X_1^4 \\otimes X_0^2, where X_i acts on the
+        i-th spin system in the register.
         """
-        if isinstance(other, numbers.Number):
-            # Create copy of the SpinOp in which every SpinOperator is multiplied by
-            # `other`.
-            new_operatorlist = [copy.deepcopy(base_operator) * other
-                                for base_operator in self.operator_list]
-            return SpinOp(new_operatorlist)
+        return self._spin_array[0]
```

Comment:
Would it make sense to create a separate class Spin that would have x, y, z fields? It would make the code more readable - here we need to know the convention that 0 is x, 1 is y etc. Usually, having data in an array and accessing it by indices is a bad pattern. Easy to make a mistake, more difficult to spot it and you can also get an index out of bounds.

Reply:
It cannot use NumPy methods such as transpose and stack so it loses the performance.
If you can make it without losing the performance, please make PR after merged.

Reply:
Ok